Reproject plugin for Imviz

pllim opened the pull request:
<!-- This comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our code of conduct,
https://github.com/spacetelescope/jdaviz/blob/main/CODE_OF_CONDUCT.md . -->

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable viz component(s). -->

This pull request is to implement a plugin to call `reproject` in Imviz, as requested by user.

Pro:

* Can run reproject from GUI (even though reproject is a completely separate package). See https://reproject.readthedocs.io/

Con:

* This plugin does not make sense for the following workflow: The data generated by reproject will align back with the original (reference) data when you link by WCS, defeating the purpose of reprojection. 
* Not performant: It took almost 30 seconds on my machine to rotate a single 2k by 4k image (the first one in `ImvizDitheredExample.ipynb`)
* See "Blocker(s)" below.

Blocker(s):

* ~Does not work on GWCS until this issue is addressed:~ https://github.com/astropy/reproject/issues/328

### TODO

- [x] What to do with original (unrotated) image?
- [x] `Out of range float values are not JSON compliant`
- [x] Let user customize new label; have default label not be so ugly.
- [x] Add tests.
- [x] Make sure change log in next release section.
- [x] Is it okay to not expose things as public API yet?
- [ ] Do we have to rename this plugin?
- [ ] Any way to make sure Compass zoom box is always correct? Seems to only happen when you reproject the smaller image and then use the reprojected smaller image as reference, and then view a larger image. We'll just open a follow-up ticket for this one to deal with later.

### Change log entry

- [x] Is a change log needed? If yes, is it added to `CHANGES.rst`? If you want to avoid merge conflicts,
  list the proposed change log here for review and add to `CHANGES.rst` before merge. If no, maintainer
  should add a `no-changelog-entry-needed` label.

### Checklist for package maintainer(s)
<!-- This section is to be filled by package maintainer(s) who will
review this pull request. -->

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [x] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [x] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [x] Is a milestone set? Set this to bugfix milestone if this is a bug fix and needs to be released ASAP; otherwise, set this to the next major release milestone.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)? [🐱](https://jira.stsci.edu/browse/JDAT-3016)

camipacifici commented:
<img width="851" alt="Screen Shot 2022-12-23 at 9 28 09 PM" src="https://user-images.githubusercontent.com/19998720/209418212-872dec57-ba27-4991-8f34-0c1b23fa686b.png">
Got this.
Looks like the meta data keyword is called "wcsinfo.wcsaxes".

camipacifici commented:
Oh, apologies. I did not see the blocker. I will try with a non-gwcs image.

camipacifici commented:
There is much more to discuss about this, but just a picky detail, the image is visible but according to the data menu it is not loaded in the viewer.

<img width="930" alt="Screen Shot 2022-12-23 at 9 41 31 PM" src="https://user-images.githubusercontent.com/19998720/209418514-b9032ddd-6ace-47c4-8cb8-8c19949950ab.png">

pllim commented:
> it is not loaded in the viewer

It should eventually autoload if you wait for it to finish. But if you do something else while it is not done, the callback might be cancelled or something, not sure (it did auto display for me). Like I said, there is a serious performance issue because reprojection is not meant to be some instant on-the-fly visualization; it is very calculation intensive (albeit an one-time overhead but after that it eats the memory).

camipacifici commented:
I played a bit with this and here are some thoughts in random order:
- I could get the plugin to work on jwst images using the approximated wcs in the header. I consider the gwcs problem external.
- The reproject call is a little slower within jdaviz than outside of jdaviz, but not dramatically slower. Performance should be investigated with multiple images or multiple calls to reproject.
- If I do the reproject outside of jdaviz, I am left with an image and a new wcs, but no combined product. I can load the image into jdaviz, but I will miss the wcs. If I want both, I need one more step to write out a file with the image and its wcs before loading into jdaviz.
- If I load the original image into jdaviz and run reproject within jdaviz, I have the combined product (image+wcs) automatically (very convenient).
- If I reproject in jdaviz, then load another image, and align by wcs everything goes back to the original orientation (very invonvenient).
- If I reproject in jdaviz, then unload the original image before loading a new image, the align by wcs works like a charm and all subsequent images are north up (or aligned as I wanted). So the overhead of reproject happens only once per workflow. If the original image could be unloaded automatically when the reproject happens, I think there would be a big gain.
- I do not know if other problems will be triggered by unloading the original image.
- If I understood correctly, there is no need for the actual reproject package because the same functionality can be reproduced with just astropy.
@orifox @astrofrog thoughts?

pllim commented:
> I consider the gwcs problem external

Sure, but Imviz loads the GWCS when it is available. To ask Imviz to use the FITS WCS version is not possible right now. To make that possible is a separate ticket.

pllim commented:
> The reproject call is a little slower within jdaviz than outside of jdaviz

How did you profile this? We have to make sure we're not comparing apples to oranges here. Overhead of GUI/linking is unavoidable in Jdaviz but that is something using `reproject` directly does not have to worry about.

pllim commented:
> If I want both, I need one more step to write out a file with the image and its wcs before loading into jdaviz

You can also avoid writing a file by creating `NDData` object with all the info within and then loading that into Jdaviz. But yes, it is one extra step (you can also see that done in the plugin code here).

pllim commented:
> there is no need for the actual reproject package because the same functionality can be reproduced with just astropy

Not sure what you mean here, can you please elaborate, @camipacifici ?

camipacifici commented:
> > I consider the gwcs problem external
> 
> Sure, but Imviz loads the GWCS when it is available. To ask Imviz to use the FITS WCS version is not possible right now. To make that possible is a separate ticket.

I was just thinking that this problem can be fixed on the reproject side so that the package does not crush with gwcs.

> > The reproject call is a little slower within jdaviz than outside of jdaviz
> 
> How did you profile this? We have to make sure we're not comparing apples to oranges here. Overhead of GUI/linking is unavoidable in Jdaviz but that is something using `reproject` directly does not have to worry about.

No profiling, just observation. Nothing major anyway.

> > If the original image could be unloaded automatically when the reproject happens
> 
> This is technically possible but it opens up a can of worms for complicated workflows.
> 
> For example, you load the original image, have markers defined by pixel against the original image, then you reproject, now if you unload the original image, what do you want to happen to those markers?
> 
> Also, at this rate, why even do reproject inside Jdaviz? I still don't see the point when you can do this all outside of Jdaviz and then load the result after.

Good point, I did not think about that. More use cases to consider here.
And the need inside the tool is for our fellow users who work in the stand alone app.

> > there is no need for the actual reproject package because the same functionality can be reproduced with just astropy
> 
> Not sure what you mean here, can you please elaborate, @camipacifici ?

Sorry, this is what I thought I understood from Tom's example reprojection. I thought he did not call the reproject package, but just the astropy routines that are used within the reproject package (the pixel-to-pixel method?).

pllim commented:
> Tom's example reprojection

Did this discussion happen while I was out? Can you please link to his example? Thanks!

pllim commented:
> users who work in the stand alone app

If we can revive that front-end canvas rotation, these users will not need the reproject plugin at all. If they insist on using reproject anyway, they will have to fire off a Python session and learn some Python. Using Python while pretending to not use Python can only go so far.

camipacifici commented:
> > Tom's example reprojection
> 
> Did this discussion happen while I was out? Can you please link to his example? Thanks!

https://gist.github.com/astrofrog/720a8ce97700915d9dc965617ebdf874

astrofrog commented:
@pllim - just FYI reproject v0.10.0 is out with the fixes to the WCS helper to work with GWCS. Let me know if you run into any issues here with this!

The gist linked above has the latest approach I worked on and that one was not abandoned due to performance but put on hold while the approach here was explored to see if it would be sufficient, and it sounds from the last SME tag-up like we might also be able to revive the image rotation in the front-end canvas which would be ideal, in which case the on-the-fly reprojection can be abandoned.

It is technically possible to use astropy without reproject to do the reprojection (coding it up by hand with pixel_to_pixel) but if reproject is going to be used anyway for the WCS helper then might as well use reproject for the reprojection too, and it allows users to potentially select different reprojection modes (interp, exact, etc).

pllim commented on jdaviz/configs/imviz/plugins/reproject/reproject.py:
```diff
@@ -0,0 +1,94 @@
+class Reproject(PluginTemplateMixin, DatasetSelectMixin):
+    template_file = __file__, "reproject.vue"
+
+    results_label = Unicode("Reprojected").tag(sync=True)
```
I thought about using the mixin class but the "add result to viewer" over there is too flexible to my liking, so I opted to keep this simple for now.

https://github.com/spacetelescope/jdaviz/blob/95c34c2e5eaffd76352cfea2e93e6501c9c433c2/jdaviz/core/template_mixin.py#L1586

kecnry replied:
what is too flexible about it?  Maybe we need a subclass that is less flexible or just options to make it do what you want here?

pllim replied:
I want to be able to control exactly which viewer it is going to and then do more stuff to that viewer, so I cannot use any of the viewer logic in the mixin but I also don't know how to turn it off over there.

kecnry replied:
If you can't set the filter to only resolve to the viewer you want, then we should add that capability so that you can.  Once that is in place, the dropdown will become a switch and you'll still get all the consistent labeling logic used in all the other plugins.

kecnry replied:
all other plugins use a switch (so that you can add the results to the data collection, but not to a viewer), but if we really don't want that here, we can add an option to hide the switch as well.  I just think we should try to use these components so that behavior does not diverge between plugins.

kecnry replied:
then let's either hide the switch, force it to be on, or use the `AutoTextField` component for the label instead and just live with some code duplication?

pllim replied:
All those SGTM, which one would have the least code impact?

kecnry replied:
After talking, I think the `AutoTextField` will be sufficient.  That will let you set the default label to be based on the input data-label (perhaps appending `(reprojected)`) and you can easily add validation errors if its invalid or overwriting.  If you want to support overwriting an existing label, then I think it makes more sense to use the `AddResults` component and am happy to help hide the switch.

bmorris3 commented on jdaviz/configs/imviz/plugins/reproject/reproject.py:
```diff
@@ -0,0 +1,94 @@
+                raise ValueError(
+                    f'{self.results_label} is already used, choose another label name.')
+
+            # Find WCS where North is pointing up.
+            wcs_out, shape_out = find_optimal_celestial_wcs([(data.shape, wcs_in)], frame='icrs')
+
+            # Reproject image to new WCS.
+            comp = data.get_component(data.main_components[0])
+            new_arr = reproject_interp((comp.data, wcs_in), wcs_out, shape_out=shape_out,
+                                       return_footprint=False)
```
Sorry if this has already been discussed and I'm repeating a question. Should we support reprojection onto non-optimal WCS? For example, one might want to rotate one image so that it matches the orientation of another (see [reproject Quick Start](https://reproject.readthedocs.io/en/stable/index.html#quick-start)), which might be helpful for blinking in Imviz.

pllim replied:
Ori wanted N-up, so that is what he is getting. Arbitrary rotation would be a different ticket.

pllim commented:
If you run this on the second image in Imviz example notebook, the input shape is `(2265, 2269)` and output shape is `(3066, 3065)`. It took a little over 8 seconds on my machine.

pllim commented on jdaviz/configs/imviz/plugins/reproject/reproject.py:
```diff
@@ -0,0 +1,94 @@
+
+
+@tray_registry('imviz-reproject', label="Reproject")
+class Reproject(PluginTemplateMixin, DatasetSelectMixin, AutoTextFieldMixin):
```
I am guessing just throwing `AutoTextFieldMixin` in there isn't enough? I am getting error about not enough arguments for the constructor. And I can't quite figure it out from the API doc. Maybe I am using this wrong, @kecnry ?

kecnry replied:
Looking into it.

kecnry commented:
Seems to work quite well!  If this is going to be limited to N up (forever... which seems unlikely), then I would advocate for merging with the compass plugin.  Otherwise, I think we need to think carefully about the name for this plugin and the future canvas rotation so that the different use-cases are as clear as possible from the names.  "Reproject" alone doesn't make me think of rotation, but we also can't call it "Rotation" if we're going to have the canvas rotation live side-by-side... 🤔  Perhaps "Rotate (reproject)" or "Rotate (data)" vs "Rotate (viewer)"?

bmorris3 commented:
> "Reproject" alone doesn't make me think of rotation, but we also can't call it "Rotation" if we're going to have the canvas rotation live side-by-side... 🤔 Perhaps "Rotate (reproject)" or "Rotate (data)" vs "Rotate (viewer)"?

Could we put all rotation methods in a "Rotate" plugin, with a button for "North up (reproject data)", and another for "Rotate (canvas)"? With enough instructive hints next to the buttons, that would work for me.

kecnry commented:
> Could we put all rotation methods in a "Rotate" plugin, with a button for "North up (reproject data)", and another for "Rotate canvas"? With enough instructive hints next to the buttons, that would work for me.

I like that idea if additional functionality won't be added (although Canvas rotation could/should also have a convenience button to rotate to the correct angle to place north up).... but I'm worried it'll get clunky otherwise.  If both were to get more complicated but in similar ways, we could also have a single plugin but with a switch (rotate data or rotate axes/viewer/canvas).  That might be cleanest for the user, but means we need to keep the features between the two approaches in parity.

pllim commented:
I have a feeling it will have additional features added in the future, so we cannot make the name too specific. I also thought about "The thing that Ori really wanted".

pllim commented:
I don't get the hesitation to name this "Reproject" though. It is basically GUI to run `reproject`. I think the name is clear enough.

bmorris3 commented:
> I don't get the hesitation to name this "Reproject" though.

Reprojection is one of several possible ways of rotating an image. If we're going to group them later, it would be strange to stick with a name that's specific to the method. `Line Analysis` isn't `Specutils`.

kecnry commented:
And "reproject" as a word can mean much more than rotation, it just feels ambiguous to me.  But if the intention is to handle distortion or various methods for projecting large images from the celestial sphere onto the screen... then keeping the general name makes sense to me.

pllim commented:
> But if the intention is to handle distortion or various methods for projecting

Distortion correct is not going to happen, but as Brett mentioned, it is possible to reproject onto another image's WCS.

For large images, it depends on how much memory you have on the machine.

kecnry commented on jdaviz/configs/imviz/plugins/reproject/reproject.py:
```diff
@@ -0,0 +1,122 @@
+                            coords=wcs_out,
+                            data=np.nan_to_num(new_arr, copy=False))
+            new_data.meta.update({'orig_label': data.label,
+                                  'reproject_version': reproject.__version__})
```
this is the only line that requires the within-method import of `reproject`... would it be cleaner to just import `__version__ as reproject_version` at the top?

pllim replied:
Done.

kecnry commented on jdaviz/configs/imviz/plugins/reproject/reproject.py:
```diff
@@ -0,0 +1,122 @@
+        data = self.data_collection[self.dataset_selected]
+        wcs_in = data.coords
+        if wcs_in is None:
+            return
```
should this raise an error or snackbar instead of failing silently?

pllim replied:
Added a snackbar message.

kecnry commented on jdaviz/configs/imviz/plugins/reproject/reproject.py:
```diff
@@ -0,0 +1,122 @@
+        self.reproject_in_progress = True
+        t_start = time.time()
+        try:
+            if self.label in self.app.data_collection.labels:
+                raise ValueError(
+                    f'{self.label} is already used, choose another label name.')
```
could now check and raise `self.label_invalid_msg` which would also handle the case of an empty label (or any other logic we add above in the future)

kecnry commented on jdaviz/configs/imviz/plugins/reproject/reproject.py:
```diff
@@ -0,0 +1,122 @@
+            viewer = self.app.get_viewer(viewer_reference)
+            viewer.state.reference_data = new_data
+
+        except Exception as e:
```
`Exception` seems quite broad to me, but I guess if there are any bugs in the block above outside the actual reprojection, CI should catch that?  Although the CI itself would probably not expose the snackbar error message, right?

pllim replied:
I cannot know all the possible ways this can crash, including invalid calculations over at reproject. CI can still check to see if the new data is generated as expected or not, even if we don't throw exception here.

kecnry replied:
ok, we've done similar elsewhere, so I think its ok.  If those CI tests ever fail, we'll just have to modify this to throw the exception in a local run of the test to see the full traceback.

pllim commented on jdaviz/core/template_mixin.py:
```diff
@@ -1567,7 +1567,7 @@ class AutoTextFieldMixin(VuetifyTemplate, HubListener):
         :default="label_default"
         :auto.sync="label_auto"
-        :invalid_msg="invalid_msg"
+        :invalid_msg="label_invalid_msg"
```
Ah, so I wasn't going crazy! Thanks for fixing it, Kyle. 😸

kecnry replied:
It was mostly my fault for having incorrect dev docs for the mixin (now fixed) 😉

pllim commented:
@kecnry or @duytnguyendtn , I vaguely remember we fixed the filename too long problem some time ago for a different plugin but it creeped back or is this something entirely different? You can reproduce this by using the Reproject plugin from this PR on the Imviz example notebook.

![Screenshot 2023-02-10 141054](https://user-images.githubusercontent.com/2090236/218177634-0fab3172-169b-475b-88cc-318c1f7d38aa.jpg)

Also, Kyle, is the result label supposed to auto-update when I select new data? I thought it is from the code but it is not updating.

![Screenshot 2023-02-10 141120](https://user-images.githubusercontent.com/2090236/218177769-de85cdb9-274e-4d0e-925c-0a98ab3217dd.jpg)

kecnry commented:
>  I vaguely remember we fixed the filename too long problem some time ago for a different plugin but it creeped back or is this something entirely different? 

hmmm, I wasn't seeing this.  Do you see it on all plugins?  Just in lab or notebook?  We've had similar things fixed upstream multiple times, but might have crept back in 😬 

> is the result label supposed to auto-update when I select new data? I thought it is from the code but it is not updating.

Once you manually change the string, then it no longer auto-updates.  Click the 🏷️ button on the right to update to the default and re-sync with changes to the dataset.  If that un-syncing is being triggered by pressing the END button but not editing the string... that is a bug that we should look into that will likely affect all label components.

duytnguyendtn commented:
I do not recall that we actually fixed this. I believe what you're referring to is the discussion regarding https://github.com/spacetelescope/jdaviz/pull/1673; In that PR, my memory is that it was decided the tradeoff of attaching the menus was worth the inconvenience of overflowing the text because the user could always "just expand the toolbar" and the cases where this would be an issue would be minimal. That being said, I also recall we would keep our eyes open in case this became a big issue down the road. Perhaps this is that point?

pllim commented:
Re: Label too long -- I don't see this problem in other plugins for the same file (e.g., Metadata). It renders differently for Reproject.

kecnry commented:
Ah, its caused by the data dropdown, not the label component - I can reproduce now.  Definitely worth looking into seeing if any of the css or vue layout is different between the two then.

pllim commented:
> If that un-syncing is being triggered by pressing the END button

Yes, just pressing END triggered it.

kecnry commented:
I wonder if wrapping in the progress spinner is the difference causing the overflow... 🤔

kecnry commented:
> Yes, just pressing END triggered it.

Ok, it might be unsyncing on any keypress.  Please file a follow-up issue/ticket for that and I think its safe to ignore for the purpose of this PR.

pllim commented:
Yes, wrapping the progress spinner is the culprit of label overflowing. It renders fine when I moved this out of the labels.

```html
      <div style="display: grid"> <!-- overlay container -->
          <div style="grid-area: 1/1">
```

pllim commented:
Okay, I think I addressed all the comments thus far, except for the plugin name that is being debated still.

Also, when I reproject the larger image in the Imviz example notebook, the image shape is `(4636, 4648)` and the reprojected shape is `(6291, 6286)`. It look 42 seconds and I could hear my CPU churning away. cc @larrybradley

kecnry commented:
Is the compass a blocker or a follow-up bug (thanks for reminding me with the commit to add the note to the docs)?

Besides that and the naming/placement, the code looks good to me!

pllim commented:
> Is the compass a blocker

Not really. It doesn't always happen. If you reproject the first (larger) image, the second image actually shows up fine on Compass, but not the other way around. I suspect this is a general Compass limitation when two images have vastly different dimensions. I can open up a follow-up ticket after this PR is merged.

pllim commented:
Cami said I should move this back to draft while Ori & co ponder.

pllim commented:
Superseded by #2179
